### PR TITLE
[new release] mirage-block-ccm (1.0.2)

### PR DESCRIPTION
**Pull request by sg2342**
CHANGES:

 * update to the current ecosystem

### CODE REVIEW

**Comment by kit-ty-kate** [comment]
```
#=== ERROR while fetching sources for mirage-block-ccm.1.0.2 ==================#
OpamSolution.Fetch_fail("https://github.com/sg2342/mirage-block-ccm/releases/download/1.0.2/mirage-block-ccm-1.0.2.tbz (curl: code 404 while downloading https://github.com/sg2342/mirage-block-ccm/releases/download/1.0.2/mirage-block-ccm-1.0.2.tbz)")
```

**Comment by sg2342** [comment]
@kit-ty-kate : sorry about that; i had `dune-release publish ...` issues.
`mirage-block-ccm-1.0.2.tbz` is now an asset of the `1.0.2` release and the fetch works.

File: packages/mirage-block-ccm/mirage-block-ccm.1.0.2/opam [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+]
+
+depends: [
+  "ocaml" {>= "4.08.0"}
+  "dune" {>= "2.7.0"}
+  "cstruct" {>= "6.0.0"}
+  "lwt" {>= "2.4.3"}
+  "mirage-block" {>= "2.0.0"}
+  "mirage-crypto"
+  "mirage-crypto-rng"
```
Comment (kit-ty-kate) [inline packages/mirage-block-ccm/mirage-block-ccm.1.0.2/opam:39]:
```suggestion
  "mirage-crypto" {>= "0.8.1"}
  "mirage-crypto-rng" {>= "0.8.1"}
```
```
#=== ERROR while compiling mirage-block-ccm.1.0.2 =============================#
# context              2.1.2 | linux/x86_64 | ocaml-base-compiler.4.10.2 | pinned(https://github.com/sg2342/mirage-block-ccm/releases/download/1.0.2/mirage-block-ccm-1.0.2.tbz)
# path                 ~/.opam/4.10/.opam-switch/build/mirage-block-ccm.1.0.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p mirage-block-ccm -j 31
# exit-code            1
# env-file             ~/.opam/log/mirage-block-ccm-6432-bead20.env
# output-file          ~/.opam/log/mirage-block-ccm-6432-bead20.out
### output ###
#       ocamlc lib/.block_ccm.objs/byte/block_ccm.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.10/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/.block_ccm.objs/byte -I /home/opam/.opam/4.10/lib/bigarray-compat -I /home/opam/.opam/4.10/lib/bytes -I /home/opam/.opam/4.10/lib/cstruct -I /home/opam/.opam/4.10/lib/fmt -I /home/opam/.opam/4.10/lib/lwt -I /home/opam/.opam/4.10/lib/mirage-block -I /home/opam/.opam/4.10/lib/mirage-crypto -I /home/opam/.opam/4.10/lib/mirage-crypto-rng -I /home/opam/.opam/4.10/lib/mirage-device -I /home/opam/.opam/4.10/lib/ocplib-endian -I /home/opam/.opam/4.10/lib/result -I /home/opam/.opam/4.10/lib/seq -intf-suffix .ml -no-alias-deps -o lib/.block_ccm.objs/byte/block_ccm.cmo -c -impl lib/block_ccm.ml)
# File "lib/block_ccm.ml", line 52, characters 12-67:
# 52 |       match Mirage_crypto.Cipher_block.AES.CCM.authenticate_decrypt ~key ~nonce ~adata c with
#                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: Unbound value Mirage_crypto.Cipher_block.AES.CCM.authenticate_decrypt
#     ocamlopt lib/.block_ccm.objs/native/block_ccm.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.10/bin/ocamlopt.opt -w -40 -g -I lib/.block_ccm.objs/byte -I lib/.block_ccm.objs/native -I /home/opam/.opam/4.10/lib/bigarray-compat -I /home/opam/.opam/4.10/lib/bytes -I /home/opam/.opam/4.10/lib/cstruct -I /home/opam/.opam/4.10/lib/fmt -I /home/opam/.opam/4.10/lib/lwt -I /home/opam/.opam/4.10/lib/mirage-block -I /home/opam/.opam/4.10/lib/mirage-crypto -I /home/opam/.opam/4.10/lib/mirage-crypto-rng -I /home/opam/.opam/4.10/lib/mirage-device -I /home/opam/.opam/4.10/lib/ocplib-endian -I /home/opam/.opam/4.10/lib/result -I /home/opam/.opam/4.10/lib/seq -intf-suffix .ml -no-alias-deps -o lib/.block_ccm.objs/native/block_ccm.cmx -c -impl lib/block_ccm.ml)
# File "lib/block_ccm.ml", line 52, characters 12-67:
# 52 |       match Mirage_crypto.Cipher_block.AES.CCM.authenticate_decrypt ~key ~nonce ~adata c with
#                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: Unbound value Mirage_crypto.Cipher_block.AES.CCM.authenticate_decrypt
```

**Comment by kit-ty-kate** [comment]
This https://github.com/sg2342/mirage-block-ccm/blob/master/examples/dune#L6 line is suspicious. I would highly suggest to avoid `(optional)` at all cost, but if you have to use it then there are missing optional dependencies in the opam file that need to be added.

**Comment by hannesm** [comment]
@kit-ty-kate the optional dependencies are marked as `dev` in the opam file, similarly to e.g. u2f and webauthn. Is there a better way to do this?

**Comment by kit-ty-kate** [comment]
What is the purpose of this tool and why would it be only installed in `dev` mode?

The issue with using `dev` is that in 99% of the case the dependencies are not defined (dev = false) and having `(optional)` means if all the libraries required are installed, the tool will be installed. Which mean now the package is non-reproducible. Worse even, if one of the dependency is in the middle of installing, dune might detect it as installed and fail halfway through because it isn’t fully installed yet.

**Comment by hannesm** [comment]
I think I understand what you mean - but maybe that should be addressed by the dune developers (why do they ship an `optional` thingy when they have seen opam's `depopts`?). For "now the package is non-reproducible" I guess my point of view is different: reproducibility means you know the set of system packages and opam packages that are installed. In any case, maybe dropping the optional and `dev` is the way to go.

**Comment by hannesm** [comment]
doesn't matter -- see https://github.com/sg2342/mirage-block-ccm/pull/25

**Comment by hannesm** [comment]
Now that https://github.com/sg2342/mirage-block-ccm/pull/25 is merged, would you could another release (or remove the git tag and release, and publish a fresh release under the same name), @sg2342? Thanks :)

**Comment by mseri** [comment]
What is the status of this? Do we need a new PR?

**Comment by mseri** [comment]
Ping @hannesm @sg2342

**Comment by hannesm** [comment]
superseeded by #22364

**Pull request closed without merging**